src/AppInstallerCommonCore/Yaml.cpp

```diff
@@ -0,0 +1,478 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include <pch.h>
+#include "winget/Yaml.h"
+#include "YamlWrapper.h"
+#include "AppInstallerErrors.h"
+#include "AppInstallerLogging.h"
+
+
+namespace AppInstaller::YAML
+{
+    using namespace std::string_view_literals;
+
+    namespace
+    {
+        Node s_globalInvalidNode;
+
+        std::string_view GetExceptionTypeStringView(Exception::Type type)
+        {
+            switch (type)
+            {
+            case Exception::Type::None:
+                return "None"sv;
+            case Exception::Type::Memory:
+                return "Memory"sv;
+            case Exception::Type::Reader:
+                return "Reader"sv;
+            case Exception::Type::Scanner:
+                return "Scanner"sv;
+            case Exception::Type::Parser:
+                return "Parser"sv;
+            case Exception::Type::Composer:
+                return "Composer"sv;
+            case Exception::Type::Writer:
+                return "Writer"sv;
+            case Exception::Type::Emitter:
+                return "Emitter"sv;
+            }
+
+            return "Unknown"sv;
+        }
+
+        void OutputExceptionHeader(std::ostringstream& out, Exception::Type type)
+        {
+            out << "[YAML:" << GetExceptionTypeStringView(type) << "] ";
+        }
+
+        void OutputMark(std::ostringstream& out, const Mark& mark)
+        {
+            out << "[line " << mark.line << "; col " << mark.column << ']';
+        }
+    }
+
+    Exception::Exception(Type type) :
+        wil::ResultException(APPINSTALLER_CLI_ERROR_LIBYAML_ERROR)
+    {
+        std::ostringstream out;
+        OutputExceptionHeader(out, type);
+
+        if (type == Type::Memory)
+        {
+            out << "Unable to (re)allocate memory";
+        }
+        else
+        {
+            out << "An unknown error occurred";
+        }
+
+        m_what = out.str();
+    }
+
+    Exception::Exception(Type type, const char* problem, size_t offset, int value) :
+        wil::ResultException(APPINSTALLER_CLI_ERROR_LIBYAML_ERROR)
+    {
+        std::ostringstream out;
+        OutputExceptionHeader(out, type);
+
+        out << (problem ? problem : "Unexplained error");

```

Comment:
>"Unexplained error" [](start = 36, length = 19)

nit; I feel "Unknown error" is less confusing for me

Reply:
After a second thought, "unexplained" seems better fit

---
In reply to: [468993477](https://github.com/microsoft/winget-cli/pull/535#discussion_r468993477) [](ancestors = 468993477)